add new metric for the internal client-go cache size

aojea opened the pull request:
/kind feature

```release-note
client-go exposes two new metrics to monitor the client-go logic that
generate http.Transports for the clients.

- rest_client_transport_cache_entries is a gauge metric
with the number of existin entries in the internal cache

- rest_client_transport_create_calls_total is a counter
that increments each time a new transport is created, storing
the result of the operation needed to generate it: hit, miss
or uncacheable
```

/sig api-machinery
/sig instrumentation

aojea commented:
/assign @enj @liggitt @logicalhan

enj commented on staging/src/k8s.io/client-go/transport/cache.go:
```diff
@@ -80,6 +81,7 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
 		// Ensure we only create a single transport for the given TLS options
 		c.mu.Lock()
 		defer c.mu.Unlock()
+		defer metrics.TransportCacheEntries.Observe(context.Background(), float64(len(c.transports)))
```
Er what exactly is the purpose of using a context if it is always set to background?

aojea replied:
copy paste thing

aojea commented on staging/src/k8s.io/client-go/tools/metrics/metrics.go:
```diff
@@ -64,6 +64,11 @@ type RetryMetric interface {
 
+// TranportCacheMetric counts the number of entries in the internal transport cache
+type TranportCacheMetric interface {
+	Observe(ctx context.Context, value float64)
```
I better ask @logicalhan and @dgrisonnet if this is a convention or just "culto cargoing" with the current state

dgrisonnet replied:
Keeping it as a float would make it consistent with Prometheus Observer interface: https://github.com/prometheus/client_golang/blob/v1.15.0/prometheus/observer.go#L18-L20

But I am also fine with simplifying the code by hiding the cast.

logicalhan replied:
Yeah I'm find with hiding the cast as well.

aojea replied:
then I remove the context too

benluddy commented:
Is it useful to add counters for cache accesses (hit and miss) too?

aojea commented:
> Is it useful to add counters for cache accesses (hit and miss) too?

/hold

that is indeed a great idea

aojea commented:
added

> > Is it useful to add counters for cache accesses (hit and miss) too?
> 
> /hold
> 
> that is indeed a great idea

/hold cancel

added Ben suggestion, in addition record no-cacheable calls too

aojea committed:
use client-go metrics not only rest_client

Change-Id: Id1a98fbaffd3e9db8dc719478ac1e7b4031ccbad

aojea committed:
register missing metrics

Change-Id: I8d00e780a86ade5ab2533df8cdf5b1feffb5c6ce

aojea commented:
> @aojea if you are feeling particularly generous you could add an integration test that [#117250 (comment)](https://github.com/kubernetes/kubernetes/issues/117250#issue-1664914729):
> 
> >

added a test, but I feel risky to make assumptions and assume that the creation of transports is a constant after the cluster is up.
There is also the problem of loading the metrics,  since it is a global variable, I think that this test gives us certain confidence thought.

I've found one bug of one metrics that was not registered, hence not reporting data and another problem with kube-proxy and kubelet not loading all the metrics.

dgrisonnet commented on staging/src/k8s.io/component-base/metrics/prometheus/restclient/metrics.go:
```diff
@@ -164,6 +182,9 @@ func init() {
 	legacyregistry.MustRegister(requestRetry)
 	legacyregistry.RawMustRegister(execPluginCertTTL)
 	legacyregistry.MustRegister(execPluginCertRotation)
+	legacyregistry.MustRegister(execPluginCalls)
```
oh wow, nice catch. We might want to implement a static analysis tool to catch those

enj replied:
Not sure how we would catch this via static analysis, but I think it would be easy enough to write a new e2e (in a separate PR) that checks the complete set of metrics registered against the API server (so that list would be updated every time someone added something new).

aojea replied:
> the complete set of metrics registered against the API server (so that list would be updated every time someone added something new)

this is a good issue for new contributors

dgrisonnet replied:
Yeah an e2e test will definitely be a good addition :+1:

I was thinking of having something static in order to catch metrics that are instantiated but not registered so that we don't have to rely on every component having up-to-date tests, although it would be great if they had.

aojea replied:
how do you do that? do you parse all files?

dgrisonnet replied:
yeah that's what I had in mind, but I don't know if it is feasible.

For instance that is already what we are doing to verify metrics stability: https://github.com/kubernetes/kubernetes/blob/master/test/instrumentation/stability-utils.sh

dgrisonnet commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1347,59 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	// TODO: this can be much better if there is some library that parse prometheus metrics
+	// the existing one in "k8s.io/component-base/metrics/testutil" uses the global variable
+	// but we want to parse the ones returned by the endpoint to be sure the metrics are
+	// exposed correctly
```
Could [ScrapeAndCompare](https://github.com/kubernetes/component-base/blob/v0.27.0/metrics/testutil/testutil.go#L70-L75) perhaps cover your needs?

dgrisonnet replied:
Sure. I just wanted to make sure this didn't solve the problem you mentioned, but I am totally fine with your current approach :+1:

aojea replied:
it was a good suggestion

dgrisonnet reviewed (approved):
/lgtm
/approve

enj commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1347,59 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+		t.Errorf("expected more hits %d in the cache than misses %d", hits, misses)
+	}
+
+	// TODO: we can keep querying for some time to see if the misses keep growing over time
```
Would it be valuable to spam ~1k updates to an API service and them assert that the total number of client has not increased more than say 100?  Basically a test that would fail if we reverted / regressed #117258 ?

aojea replied:
I like that

aojea replied:
done, aiming for a 90% hit ratio in the cache, that I think is reasonable

enj commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1347,59 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
 	}
 }
+
+func TestAPIServerTransportMetrics(t *testing.T) {
```
Probably good to enable everything:

```go
	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, "AllAlpha", true)()
	defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, "AllBeta", true)()
```

aojea replied:
good idea

enj commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1347,59 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+
+	c := clientset.NewForConfigOrDie(result.ClientConfig).RESTClient()
+
+	body, err := c.Get().AbsPath("/metrics").DoRaw(context.Background())
```
Before checking the metrics it would probably be good to create one of every resource via `GetEtcdStorageData` to exercise all the common code paths.

aojea replied:
copied pasted from other test, not influence at all, that is good

aojea commented on test/integration/client/client_test.go:
```diff
@@ -53,6 +54,10 @@ import (
+
+	// the metrics are loaded on cmd/kube-apiserver/apiserver.go
+	// so we need to load them here to be available for the test
+	_ "k8s.io/component-base/metrics/prometheus/restclient"
```
this is the problem with metrics, that the are global, so all the tests that run in the same test binary access the same metrics registry.
Since there are other test with metrics I will enable only them here by now, I do not want to do this global import because of the risk of impacting other tests , and then not catch it on presubmit, flakes pointing to this pr, revert, ...

enj commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1355,192 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+
+	for _, resourceList := range resourceLists {
+		for _, resource := range resourceList.APIResources {
+			if !strings.HasSuffix(resource.Name, "/status") {
```
Is this backwards?

```suggestion
			if strings.HasSuffix(resource.Name, "/status") {
```

aojea replied:
I copied from one test, it seems that does check status only, good catch

aojea replied:
if I change this line the test fails, I need to understand the logic it uses now 🙃

aojea replied:
I'm not sure this really makes a difference, I already have a test that verifies the client-go generated clientset shares a transport

enj replied:
I am not sure I understand - why does the test fail?  It is not clear to me why you need the if statement at all?

enj replied:
Question: would it make the metrics results easier to reason about if you did:

1. Create/delete everything
2. Global metrics delete via the equivalent REST call `kubectl delete --raw /metrics`
3. Create/delete everything again

At that point the client numbers should be stable enough to require like a 99% cache hit rate?

This may be getting too complex so I am fine with whatever you think is sufficient.

aojea replied:
I don't see any difference with and without this code, the apiserver--etcd clients are not using this cache, and the client -- apiserver are already tested, since we changed the generator to share the transport for the generated clients, I'm going to remove this code, is complex to understand and makes the test harder to rread

aojea replied:
oh, I replied and I didn't see your answers

> I am not sure I understand - why does the test fail? It is not clear to me why you need the if statement at all?

I don't know, if I change that it fails, but this code is copied verbatim from 
https://github.com/kubernetes/kubernetes/blob/6e541a6da7e1731730c71d1e723efbca5025330c/test/integration/apiserver/apply/status_test.go#L125-L129

so it may be possible there are some assumptions about the status

> Global metrics delete via the equivalent REST call kubectl delete --raw /metrics
> At that point the client numbers should be stable enough to require like a 99% cache hit rate?

there are a lot of controllers working on the background and I don't want to make assumptions that can generate flakiness in the future,

enj replied:
Note that KCM isn't running during integration tests so only a small subset of containers are running.  But yeah it's fine to drop unnecessary complexity.

enj commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1355,192 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = aggregatorClient.ApiregistrationV1().APIServices().Update(context.Background(), apiService, metav1.UpdateOptions{})
```
Can you mutate a label on this to make sure the update isn't short circuited?

aojea replied:
yeah, it now makes a difference, previously was not creating new transports, now

```
    client_test.go:1513: metric rest_client_transport_cache_entries 4
    client_test.go:1513: metric rest_client_transport_create_calls_total{result="hit"} 2063
    client_test.go:1513: metric rest_client_transport_create_calls_total{result="miss"} 4
```

enj commented:
/lgtm

aojea commented:
I tested reverting  https://github.com/kubernetes/kubernetes/pull/117258  and it fails at the beginning because of the cache/hit ratio to be lower than 90% that is ok, but if I also relax the hit ratio to 75%, it correctly detects the leak
```
go test -count 1 -timeout 30s -run ^TestAPIServerTransportMetrics$ k8s.io/kubernetes/test/integration/client -v 2>&1 | grep metric
    client_test.go:1447: metric rest_client_transport_cache_entries 3
    client_test.go:1447: metric rest_client_transport_create_calls_total{result="hit"} 23
    client_test.go:1447: metric rest_client_transport_create_calls_total{result="miss"} 3
...
    client_test.go:1447: metric rest_client_transport_cache_entries 1000
    client_test.go:1447: metric rest_client_transport_create_calls_total{result="hit"} 24
    client_test.go:1447: metric rest_client_transport_create_calls_total{result="miss"} 1001
    client_test.go:1451: expected as many entries 1000 in the cache as misses, got 1001
    client_test.go:1455: expected more hits 24 in the cache than misses 1001
    client_test.go:1403: possible transport leak, number of cache entries increased in 997
```

aojea commented:
/assign @liggitt 

if he is not already flying 🛬

enj commented:
> it correctly detects the leak

Nice, thanks for checking.

bart0sh commented:
/triage accepted
/priority important-longterm

liggitt commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1350,109 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+		t.Fatal(err)
+	}
+
+	for i := 0; i < 1000; i++ {
+		apiService, err := aggregatorClient.ApiregistrationV1().APIServices().Get(context.Background(), "v1alpha1.wardle.example.com", metav1.GetOptions{})
```
1000 writes seems high for something that isn't a stress test... can we get the same coverage more cheaply?

aojea replied:
yeah, I will try to adjust

aojea replied:
ok, I do 100 only and I verified it catches the original bug with the aggregator

aojea commented on staging/src/k8s.io/client-go/transport/cache.go:
```diff
@@ -80,11 +81,16 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
 		}
+		metrics.TransportCreateCalls.Increment("miss")
+	} else {
+		metrics.TransportCreateCalls.Increment("not-cacheable")
```
whatever is more correct in english, I do prefer a single word without hyphen

liggitt commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1350,112 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+	hits1, misses1, entries1 := checkTransportMetrics(t, client)
+	// hit ratio at startup depends on multiple factors
+	// be conservative and expect only a 75 percent
+	if (hits1*100)/(hits1+misses1) < 75 {
```
expecting repeated construction during startup seems weird... wouldn't an ideal startup construct clients exactly once, and get a hit rate of 0?

where is the hit rate currently?

aojea replied:
there are multiple process building clients, and some of them like the one this covers the regression were not cacheable ... also this runs with a default configuration, I can imaging that enabling new features may enable new controllers that build their own clientsets

The rate is ~ 88 IIRC

liggitt replied:
I'm wondering if resetting metrics at the start of the test would improve what we see here

aojea replied:
I grepped around and it seems there is no reset for metrics on the integration tests
https://github.com/kubernetes/kubernetes/blob/4cff1311c5543af30a3cbb3fa873b3fadf9ccf7a/test/integration/metrics/metrics_test.go

but I would add it so we always reset the global registry at the start of the test

aojea replied:
awesome 
```
	// reset default registry metrics
	legacyregistry.Reset()
```
actually resets all the metrics

liggitt commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1350,112 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+		// mutate the object
+		apiService.Labels = map[string]string{"key": fmt.Sprintf("val%d", i)}
+		_, err = aggregatorClient.ApiregistrationV1().APIServices().Update(context.Background(), apiService, metav1.UpdateOptions{})
+		if err != nil && !apierrors.IsConflict(err) {
```
are we concerned at all about timeouts or network flakes in the series of 100 update requests?

aojea replied:
that will be symptom that something else wrong ... also, we need to have some "large numbers" to get closer to the average, if this fails and we swallow it we may have false positives or negatives

liggitt replied:
did we run this repeatedly in CI to make sure this doesn't flake?

aojea replied:
reduced the number of requests and adding a check to avoid false positives
 run with stress
```
4m55s: 1831 runs so far, 0 failures
5m0s: 1868 runs so far, 0 failures
5m5s: 1898 runs so far, 0 failures
5m10s: 1928 runs so far, 0 failures
5m15s: 1965 runs so far, 0 failures
5m20s: 1995 runs so far, 0 failures
```

liggitt commented on test/integration/client/client_test.go:
```diff
@@ -1342,3 +1350,112 @@ func TestExtractModifyApply_ForceOwnership(t *testing.T) {
+
+	client := clientset.NewForConfigOrDie(result.ClientConfig)
+
+	hits1, misses1, entries1 := checkTransportMetrics(t, client)
```
does this see hits/misses/entry counts from all tests that ran in this package, or just this test instance?

aojea replied:
this is an interesting question, it performs a curl to the apiserver to not use the global registry

> 	body, err := client.RESTClient().Get().AbsPath("/metrics").DoRaw(context.Background())

but since all will be compiled in the same binary I assume that all apiservers running will share the same registry for metrics ... more reasons to operate on statistics and avoid running test in parallel

liggitt replied:
can we at least reset the metrics at the start of the test, to avoid other tests in the package impacting what we see?

aojea replied:
done

aojea commented:
ok, Jordan is right ... again 😬 
```
    client_test.go:1421: metric rest_client_transport_cache_entries 32
    client_test.go:1421: metric rest_client_transport_create_calls_total{result="hit"} 113
    client_test.go:1421: metric rest_client_transport_create_calls_total{result="miss"} 3
    client_test.go:1421: expected as many entries 32 in the cache as misses, got 3
```

most probably is racing with other test 🤔

logicalhan commented:
> most probably is racing with other test 🤔

Yeah there are usually two ways around this. The first is to call `Reset` at the beginning of the test. The second (slightly more involved), is to register the metric to a non-global registry (one that's isolated to your test).

aojea committed:
add new metrics for the internal client-go transport generator

Add two new metrics to monitor the client-go logic that
generate http.Transports for the clients.

- rest_client_transport_cache_entries is a gauge metrics
with the number of existin entries in the internal cache

- rest_client_transport_create_calls_total is a counter
that increments each time a new transport is created, storing
the result of the operation needed to generate it: hit, miss
or uncacheable

Change-Id: I2d8bde25281153d8f8e8faa249385edde3c1cb39

dims commented:
@aojea you doubted Jordan?!!

aojea commented:
> Kubernetes e2e suite: [It] [sig-node] Pods should run through the lifecycle of Pods and PodStatus [Conformance] expand_less	

unrelated 
/test pull-kubernetes-conformance-kind-ga-only-parallel

>  @aojea you doubted Jordan?!!

lol, I never doubted it 

> Yeah there are usually two ways around this. The first is to call Reset at the beginning of the test

I'm doing reset and moving the test to its own folder dedicated only to metrics, so we control the type of tests that are going to run together
